apps/src/javalab/javalabRedux.js

```diff
@@ -57,23 +50,36 @@ export default function reducer(state = initialState, action) {
       consoleLogs: [...state.consoleLogs, action.log]
     };
   }
-  if (action.type === SET_EDITOR_TEXT) {
-    return {
-      ...state,
-      editorText: action.editorText,
-      filesChanged: true
+  if (action.type === SET_SOURCE) {
+    let newSources = {...state.sources};
+    newSources[action.filename] = {
+      text: action.source,
+      visible: action.isVisible
     };
-  }
-  if (action.type === SET_FILENAME) {
     return {
       ...state,
-      filename: action.filename
+      sources: newSources
     };
   }
-  if (action.type === SET_FILES_CHANGED) {
+  if (action.type === RENAME_FILE) {
+    const source = state.sources[action.oldFilename];
+    if (source !== undefined) {
+      let newSources = {...state.sources};
+      delete newSources[action.oldFilename];
+      newSources[action.newFilename] = source;
+      return {
+        ...state,
+        sources: newSources
+      };
+    } else {
+      // if old filename doesn't exist, can't do a rename
```

Comment:
Is there an occasion where this would happen?

Reply:
if you accidentally sent an event twice it could happen, not sure how likely this is